src/protocol/sort/bit_permutations.rs

```diff
@@ -0,0 +1,154 @@
+            .enumerate()
+            .map(|(index, input)| async move {
+                self.context
+                    .multiply(
+                        RecordId::from(index as u32),
+                        IPAProtocolStep::Sort(SortStep::BitPermutations),
+                    )
+                    .await
+                    .execute(input.0, input.1)
+                    .await
+            });
+
+        join_all(async_multiply.map(|handle| async { handle.await.unwrap() })).await
+    }
+
+    /// Executes sorting of a bit column on mpc helpers. Each helper receives their input shares and do following steps
+    /// 1. local computation by `prepare_mult_inputs` which outputs 2 vectors [x,y]
+    /// 2. multiply each row of previous output individually (i.e. x*y) across mpc helpers.
+    /// 3. add ith column by i+len to obtain helper's share of sorted location, where len is same as input shares length
+    #[allow(dead_code)]
+    pub async fn execute<F>(self, shares: &[Replicated<F>]) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        F: Field,
+    {
+        let mult_output = self.secure_multiply(self.prepare_mult_inputs(shares)).await;
+
+        assert_eq!(mult_output.len(), shares.len() * 2);
+        // Generate permutation location
+        let len = mult_output.len() / 2;
+        Ok((0..len)
+            .map(|j| mult_output[j] + mult_output[j + len])
+            .collect::<Vec<_>>())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::rngs::mock::StepRng;
+    use tokio::try_join;
+
+    use crate::{
+        field::Fp31,
+        helpers::{self, mock::TestWorld},
+        protocol::{securemul, sort::generate_bit_permutations, IPAProtocolStep, QueryId},
+        secret_sharing::Replicated,
+    };
+
+    #[tokio::test]
+    pub async fn bit_permutations() {
+        let world: TestWorld<IPAProtocolStep> = helpers::mock::make_world(QueryId);
+        let participants = crate::helpers::prss::test::make_three();
+        let context = securemul::tests::make_context(&world, &participants);
+        let mut rand = StepRng::new(100, 1);
+        let input: Vec<u128> = vec![1, 0, 1, 0, 0, 1, 0];
+        let input_len = input.len();
+        let mut shares = [
+            Vec::with_capacity(input_len),
+            Vec::with_capacity(input_len),
+            Vec::with_capacity(input_len),
+        ];
+        for iter in input {
+            let share = securemul::tests::share(Fp31::from(iter), &mut rand);
+            for i in 0..3 {
+                shares[i].push(share[i]);
+            }
+        }
+
+        let h0_future = generate_bit_permutations(&context[0]).execute(&shares[0]);
+        let h1_future = generate_bit_permutations(&context[1]).execute(&shares[1]);
+        let h2_future = generate_bit_permutations(&context[2]).execute(&shares[2]);
+
+        let result = try_join!(h0_future, h1_future, h2_future).unwrap();
+        let expected_sort_output = [5_u128, 1, 6, 2, 3, 7, 4].as_ref();
```

Comment:
can you add an explanation why this output is expected? 

```suggestion
        let expected_sort_output = [5_u128, 1, 6, 2, 3, 7, 4];
```